jib-core/src/main/java/com/google/cloud/tools/jib/api/JavaContainerBuilder.java

```diff
@@ -0,0 +1,291 @@
+   * @return a new {@link JavaContainerBuilder} with the specified base image reference
+   * @throws InvalidImageReferenceException if {@code baseImageReference} is invalid
+   */
+  public static JavaContainerBuilder from(String baseImageReference)
+      throws InvalidImageReferenceException {
+    return from(RegistryImage.named(baseImageReference));
+  }
+
+  /**
+   * Creates a new {@link JavaContainerBuilder} with the specified base image reference.
+   *
+   * @param baseImageReference the base image reference
+   * @return a new {@link JavaContainerBuilder} with the specified base image reference
+   */
+  public static JavaContainerBuilder from(ImageReference baseImageReference) {
+    return from(RegistryImage.named(baseImageReference));
+  }
+
+  /**
+   * Creates a new {@link JavaContainerBuilder} with the specified base image reference.
+   *
+   * @param baseImageReference the base image reference
+   * @return a new {@link JavaContainerBuilder} with the specified base image reference
+   */
+  public static JavaContainerBuilder from(RegistryImage baseImageReference) {
+    return new JavaContainerBuilder(Jib.from(baseImageReference));
+  }
+
+  private JibContainerBuilder jibContainerBuilder;
+  private JavaLayerConfigurations.Builder layerConfigurationsBuilder;
+  private List<String> classpath;
+  private List<String> jvmFlags;
+  @Nullable private String mainClass;
+
+  private JavaContainerBuilder(JibContainerBuilder jibContainerBuilder) {
+    this.jibContainerBuilder = jibContainerBuilder;
+    layerConfigurationsBuilder = JavaLayerConfigurations.builder();
+    classpath = new ArrayList<>();
+    jvmFlags = new ArrayList<>();
+  }
+
+  /**
+   * Adds dependencies to /app/libs on the image.
+   *
+   * @param dependencyFiles the list of dependencies to add to the image
+   * @return this
+   * @throws IOException if adding the layer fails
+   */
+  public JavaContainerBuilder addDependencies(List<Path> dependencyFiles) throws IOException {
+    for (Path file : dependencyFiles) {
+      if (!Files.exists(file)) {
+        continue;
+      }
+      if (Files.isDirectory(file)) {
+        layerConfigurationsBuilder.addDirectoryContents(
+            LayerType.DEPENDENCIES,
+            file,
+            path -> !path.getFileName().toString().contains("SNAPSHOT"),
+            DEPENDENCIES_PATH);
+        layerConfigurationsBuilder.addDirectoryContents(
+            LayerType.SNAPSHOT_DEPENDENCIES,
+            file,
+            path -> path.getFileName().toString().contains("SNAPSHOT"),
+            DEPENDENCIES_PATH);
+      } else {
+        layerConfigurationsBuilder.addFile(
+            file.getFileName().toString().contains("SNAPSHOT")
+                ? LayerType.SNAPSHOT_DEPENDENCIES
+                : LayerType.DEPENDENCIES,
+            file,
+            DEPENDENCIES_PATH.resolve(file.getFileName()));
+      }
+    }
+    if (!classpath.contains(DEPENDENCIES_PATH.resolve("*").toString())) {
```

Comment:
Hmm, do we want to maintain the order the user sets the classpath in?

Reply:
We probably would want to maintain a defined order that doesn't depend on the order in which the user calls the builder methods.

Reply:
Could we use TreeSet? We'd probably have to implement compareTo or some kind of comparable.

Reply:
oh just kidding, you actually want it in the fixed order the user defined.

Reply:
linked hash set?

Reply:
Do we actually want it to be ordered the way the user sets it, or should it be a specific order? User-defined order makes the most sense to me but I don't know how useful that is.

Reply:
I think either works depending on what we want to guarantee. User-defined order would just mean calling builder methods in different order produces different image (which may be find since `addJvmFlag` would be that case unless we sorted those or something).

Reply:
And since it's a specific builder, we can be opinionated about how we think java should be ordered?

Reply:
I think I confused myself, but does the order of setting these layers affect how the entry point is built?